datasource/src/test/scala/org/locationtech/rasterframes/datasource/geotiff/GeoTiffDataSourceSpec.scala

```diff
@@ -101,6 +103,104 @@ class GeoTiffDataSourceSpec
       }
     }
 
+    it("should write unstructured raster") {
+      import spark.implicits._
+      val df = spark.read.format("raster")
+        .option("tileDimensions", "32,32")  // oddball
+        .load(nonCogPath.toASCIIString) // core L8-B8-Robinson-IL.tiff
+
+      df.count() should be > 0L
+
+      val crs = df.select(rf_crs($"proj_raster")).first()
+
+      noException shouldBe thrownBy {
+        df.write.geotiff.withCRS(crs).save("unstructured.tif")
+      }
+
+      val (inCols, inRows) = {
+        val id = sampleGeoTiff.imageData // inshallah same as nonCogPath
+        (id.cols, id.rows)
+      }
+      inCols should be (774)
+      inRows should be (500) //from gdalinfo
+
+      val outputTif = SinglebandGeoTiff("unstructured.tif")
+      outputTif.imageData.cols should be (inCols)
+      outputTif.imageData.rows should be (inRows)
+
+      // TODO check datatype, extent.
+    }
+
+    it("should round trip unstructured raster from COG"){
+      val df = spark.read.format("raster")
+        .load(getClass.getResource("/LC08_B7_Memphis_COG.tiff").toURI.toASCIIString())
+
+      df.count() should be > 0L
+
+      val crs = df.select(rf_crs(col("proj_raster"))).first()
+
+      noException shouldBe thrownBy {
+        df.write.geotiff.withCRS(crs).save("target/unstructured_cog.tif")
+      }
+
+      val (inCols, inRows, inExtent, inCellType) = {
+        val tif = readSingleband("LC08_B7_Memphis_COG.tiff")
+        val id = tif.imageData
+        (id.cols, id.rows, tif.extent, tif.cellType)
+      }
+      inCols should be (963)
+      inRows should be (754) //from gdalinfo
+      inExtent should be (Extent(752325.0, 3872685.0, 781215.0, 3895305.0))
+
+      val outputTif = SinglebandGeoTiff("target/unstructured_cog.tif")
+      outputTif.imageData.cols should be (inCols)
+      outputTif.imageData.rows should be (inRows)
+      outputTif.extent should be (inExtent)
+      outputTif.cellType should be (inCellType)
```

Comment:
The GeoTrellis `testkit` includes `RasterMatchers` that maybe we could use to compare the actual values?

Reply:
Ah who knew! I looked around and they seemed mostly about making sure cell values across a tile match some condition, as opposed to these kinds of metadata equalities.